python/extensions.bzl

```diff
@@ -127,3 +133,89 @@ pip = module_extension(
         "parse": tag_class(attrs = _pip_parse_ext_attrs()),
     },
 )
+
+# This function allows us to build the label name of a label
+# that is not passed into the current context.
+# The module_label is the key element that is passed in.
+# This value provides the root location of the labels
+# See https://bazel.build/external/extension#repository_names_and_visibility
+def _repo_mapped_label(module_label, extension_name, apparent):
+    """Construct a canonical repo label accounting for repo mapping.
+
+    Args:
+        module_label: Label object of the module hosting the extension; see
+          "_module" implicit attribute.
+        extension_name: str, name of the extension that created the repo in `apparent`.
+        apparent: str, a repo-qualified target string, but without the "@". e.g.
+          "python38_x86_linux//:python". The repo name should use the apparent
+          name used by the extension named by `ext_name` (i.e. the value of the
+          `name` arg the extension passes to repository rules)
+    """
+    return Label("@@{module}~{extension_name}~{apparent}".format(
```

Comment:
You may be able to use `str(Label(apparent))` to get the canonical name. It's probably not a good idea to hardcode the format of a canonical name.

cc @fmeum

Reply:
Label(apparent) won't work in this context because the repo being referenced (the platform-specific one) isn't visible (nobody called use_repo() on it).

While I agree that manually creating the canonical name isn't ideal, the other option, a hub-spoke model, requires Chris to essentially start over, which I'm not going to ask of him after a week of effort and now that we have something that works.

Reply:
@rickylev Repos generated by the same extension always see each other under the name given in the extension impl function, so the hub repo introduced in this PR should be able to just do `str(Label(apparent))` without extra complexity or `use_repo`s.

Reply:
Thanks for trying, Chris.

@fmeum @linzhp I agree with your reasoning, but it just doesn't seem to actually be working like that. This is queued for merging, so should land in a few minutes, so you'll be able to play with it and maybe be able tell us what's going on -- both Chris and I have hit a will trying to figure out what's going on.